src/services/logger.ts

```diff
@@ -18,49 +18,127 @@ if (config.nodeEnv !== 'dev' && config.nodeEnv !== 'test') {
+  #info: (message: string, serviceOperation?: string) => void = () => null;
+  #debug: (message: string, serviceOperation?: string) => void = () => null;
+  #trace: (message: string, serviceOperation?: string) => void = () => null;
+  #warn: (message: string, serviceOperation?: string) => void = () => null;
+  #error: (message: string | Error, innerError?: unknown, serviceOperation?: string) => void = () => null;
+  constructor() {
+    // enable logging only if we're on a dev environment or apm logging is activated
+    if (config.apmLogging || devEnv) {
+      switch (config.logger.logstashLevel) {
+        // error > warn > info > debug > trace
+        case 'trace':
+          this.#trace = this.#createLogCallback('trace');
+          this.#debug = this.#createLogCallback('debug');
+          this.#info = this.#createLogCallback('info');
+          this.#warn = this.#createLogCallback('warn');
+          this.#error = this.#createErrorFn();
+          break;
+        case 'debug':
+          this.#debug = this.#createLogCallback('debug');
+          this.#info = this.#createLogCallback('info');
+          this.#warn = this.#createLogCallback('warn');
+          this.#error = this.#createErrorFn();
+          break;
+        case 'info':
+          this.#info = this.#createLogCallback('info');
+          this.#warn = this.#createLogCallback('warn');
+          this.#error = this.#createErrorFn();
+          break;
+        case 'warn':
+          this.#warn = this.#createLogCallback('warn');
+          this.#error = this.#createErrorFn();
+          break;
+        case 'error':
+          this.#error = this.#createErrorFn();
+          break;
+        default:
+          break;
+      }
+    }
+  }
+
   timeStamp(): string {
-    const dateObj = new Date();
+    return createTimeStamp();
+  }
 
-    let date = dateObj.toISOString();
-    date = date.substring(0, date.indexOf('T'));
+  #createErrorFn(): (message: string | Error, innerError?: unknown, serviceOperation?: string) => void {
+    return (message: string | Error, innerError?: unknown, serviceOperation?: string): void => {
+      let errMessage = typeof message === 'string' ? message : message.stack ?? message.message;
 
-    const time = dateObj.toLocaleTimeString([], { hour12: false });
+      if (innerError) {
+        if (innerError instanceof Error) {
+          errMessage = `${errMessage}\r\n${innerError.stack ?? innerError.message}`;
+        }
+      }
+
+      logger.error(`${this.messageStamp(serviceOperation)}[ERROR] - ${errMessage}`);
+    };
+  }
 
-    return `${date} ${time}`;
+  #createLogCallback(level: 'trace' | 'info' | 'warn' | 'debug'): (message: string, serviceOperation?: string) => void {
+    switch (level) {
+      case 'trace':
+        return (message: string, serviceOperation?: string): void => {
+          logger.trace(`${this.messageStamp(serviceOperation)}[${level.toUpperCase()}] - ${message}`);
```

Comment:
instead of having "level.toUpperCase()" be called for every log in the system, having the hardcoded string "TRACE|
DEBUG|INFO" etc would be less processing.

Reply:
Good catch, no allocations needed